langchain/chains/conversation/memory.py

```diff
@@ -40,6 +46,88 @@ def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None
+    dynamic_key_list: List[str] = ["history", "entities"]  #: :meta private:
+    entities_per_turn: List[str] = Field(default_factory=list)  #: :meta private:
+
+    @property
+    def dynamic_keys(self) -> List[str]:
+        """Will always return list of dynamic keys.
+
+        :meta private:
+        """
+        return self.dynamic_key_list
+
+    def _format_buffer_to_string(self) -> str:
+        section = self.buffer[-self.history_window_len :]
+        result = []
+        for line in section:
+            result.append("Human: {input}".format(input=line["input"]))
+            result.append("AI: {response}".format(response=line["response"]))
+        return "\n".join(result)
+
+    def _load_dynamic_keys(self, inputs: Dict[str, Any]) -> Dict[str, str]:
+        """Return history buffer with entity summaries."""
+        prompt_input_keys = list(set(inputs.keys()) - set(self.dynamic_keys))
+        assert len(prompt_input_keys) == 1, "One input key expected"
+        input = inputs[prompt_input_keys[0]]
+        full_conversation = self._format_buffer_to_string() + "\nHuman: " + input
+        chain = LLMChain(llm=self.llm, prompt=self.entity_extraction_prompt)
+        entities = chain.predict(history=full_conversation)
+        if not entities:
+            self.entities_per_turn = []
+            all_results = ""
+        else:
+            all_results = ""
+            all_entities = []
+            for entity in entities.split(","):
+                entity = entity.strip()
+                all_entities.append(entity)
+                entity_summary = self.docstore.search(entity)
+                if isinstance(entity_summary, str):
+                    continue
+                all_results += "\nHere's what we know about {entity}".format(
+                    entity=entity
+                )
+                all_results += "\n" + entity_summary.page_content
+            self.entities_per_turn = all_entities
+        return {"entities": all_results, "history": self._format_buffer_to_string()}
+
+    def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None:
+        """Save context from this conversation to buffer."""
+        prompt_input_keys = list(set(inputs.keys()) - set(self.dynamic_keys))
+        assert len(prompt_input_keys) == 1, "One input key expected"
+        assert len(outputs.keys()) == 1, "One output key expected"
+        assert self.entity_summarization_prompt.input_variables == [
+            "history",
+            "entity",
+            "input",
+            "summary",
+        ], "Expected different input keys"
+        human = inputs[prompt_input_keys[0]]
+        ai = outputs[list(outputs.keys())[0]]
+        full_conversation = self._format_buffer_to_string() + "\nHuman: " + human
+        for entity in self.entities_per_turn:
+            chain = LLMChain(llm=self.llm, prompt=self.entity_summarization_prompt)
+            entity_summary = self.docstore.search(entity) or ""
```

Comment:
wait doesnt search return a document object?

Reply:
InMemoryDocstore at the moment returns a string response if the document isnt found (None might be better tbh)